flang/test/Fir/abstract-results.fir

```diff
@@ -0,0 +1,230 @@
+  %c50 = constant 50 : index
+  %buffer = fir.alloca !fir.array<?x!fir.char<1,?>>, %c100, %c50
+  %shape = fir.shape %c100 : (index) -> !fir.shape<1>
+  %res = fir.call @chararrayfunc(%c100, %c50) : (index, index) -> !fir.array<?x!fir.char<1,?>>
+  fir.save_result %res to %buffer(%shape) typeparams %c50 : !fir.array<?x!fir.char<1,?>>, !fir.ref<!fir.array<?x!fir.char<1,?>>>, !fir.shape<1>, index
+  return
+
+  // CHECK: %[[c100:.*]] = constant 100 : index
+  // CHECK: %[[c50:.*]] = constant 50 : index
+  // CHECK: %[[buffer:.*]] = fir.alloca !fir.array<?x!fir.char<1,?>>, %[[c100]], %[[c50]]
+  // CHECK: fir.call @chararrayfunc(%[[buffer]], %[[c100]], %[[c50]]) : (!fir.ref<!fir.array<?x!fir.char<1,?>>>, index, index) -> ()
+  // CHECK-NOT: fir.save_result
+
+  // CHECK-BOX: %[[c100:.*]] = constant 100 : index
+  // CHECK-BOX: %[[c50:.*]] = constant 50 : index
+  // CHECK-BOX: %[[buffer:.*]] = fir.alloca !fir.array<?x!fir.char<1,?>>, %[[c100]], %[[c50]]
+  // CHECK-BOX: %[[shape:.*]] = fir.shape %[[c100]] : (index) -> !fir.shape<1>
+  // CHECK-BOX: %[[box:.*]] = fir.embox %[[buffer]](%[[shape]]) typeparams %[[c50]] : (!fir.ref<!fir.array<?x!fir.char<1,?>>>, !fir.shape<1>, index) -> !fir.box<!fir.array<?x!fir.char<1,?>>>
+  // CHECK-BOX: fir.call @chararrayfunc(%[[box]], %[[c100]], %[[c50]]) : (!fir.box<!fir.array<?x!fir.char<1,?>>>, index, index) -> ()
+  // CHECK-BOX-NOT: fir.save_result
+}
+
+// ------------------------ Test fir.address_of rewrite ------------------------
+
+func private @takesfuncarray((i32) -> !fir.array<?xf32>)
+
+// CHECK-LABEL: func @test_address_of() {
+// CHECK-BOX-LABEL: func @test_address_of() {
+func @test_address_of() {
+  %0 = fir.address_of(@arrayfunc) : (i32) -> !fir.array<?xf32>
+  fir.call @takesfuncarray(%0) : ((i32) -> !fir.array<?xf32>) -> ()
+  return
+
+  // CHECK: %[[addrOf:.*]] = fir.address_of(@arrayfunc) : (!fir.ref<!fir.array<?xf32>>, i32) -> ()
+  // CHECK: %[[conv:.*]] = fir.convert %[[addrOf]] : ((!fir.ref<!fir.array<?xf32>>, i32) -> ()) -> ((i32) -> !fir.array<?xf32>)
+  // CHECK: fir.call @takesfuncarray(%[[conv]]) : ((i32) -> !fir.array<?xf32>) -> ()
+
+  // CHECK-BOX: %[[addrOf:.*]] = fir.address_of(@arrayfunc) : (!fir.box<!fir.array<?xf32>>, i32) -> ()
+  // CHECK-BOX: %[[conv:.*]] = fir.convert %[[addrOf]] : ((!fir.box<!fir.array<?xf32>>, i32) -> ()) -> ((i32) -> !fir.array<?xf32>)
+  // CHECK-BOX: fir.call @takesfuncarray(%[[conv]]) : ((i32) -> !fir.array<?xf32>) -> ()
+
+}
+
+// ----------------------- Test indirect calls rewrite ------------------------
+
+// CHECK-LABEL: func @test_indirect_calls(
+// CHECK-SAME: %[[arg0:.*]]: () -> ()) {
+// CHECK-BOX-LABEL: func @test_indirect_calls(
+// CHECK-BOX-SAME: %[[arg0:.*]]: () -> ()) {
+func @test_indirect_calls(%arg0: () -> ()) {
+  %c100 = constant 100 : index
+  %buffer = fir.alloca !fir.array<?xf32>, %c100
+  %shape = fir.shape %c100 : (index) -> !fir.shape<1>
+  %0 = fir.convert %arg0 : (() -> ()) -> ((index) -> !fir.array<?xf32>)
```

Comment:
I did not find a bug, but foresee troubles in FIR doing so. My issue is that I am not sure we can easily control what operations may involve function pointers in a "transparent way" (meaning that the fact that it is a function pointer does not really matter, but given the operation is typed, if we change the operand type, we will have issues).

For instance, one can imagine writing FIR using fir.alloca, fir.allocamem, fir.load, fir.store, fir.select/mlir.select, fir.coordinate_of, fir.array_coor.... involving function pointers. We could start trying to fix the types on all these ops in the rewrite pass, but it seems hard to enforce/require that future transparent ops will be added to this pass.

Here is an example of program where starting to convert the function argument signature would be problematic:

```
func @takesfuncarray(%arg0: () -> !fir.array<?xf32>) {
  %buffer = fir.alloca () -> !fir.array<?xf32>
  fir.store %arg0 to %buffer : !fir.ref<() -> !fir.array<?xf32>>
  // later ... fir.load, fir.select, and fir.call
  return
}
```

So rather than to make the conversion pass more complex, I decided to push the function pointer conversions to the edge operations where we have a strong control on: address_of and fir.call, and to not try to fix types where a function pointer might simply transit, which includes function pointer arguments.

Reply:
These are good arguments. It sounds like there is "a bug" potentially here if we're saying we're not fully rewriting the function signatures. (There is no way to enforce that your example can't be written, right?) I'm wondering if maybe we need to visit all Ops in the module and rewrite FunctionTypes everywhere.

Reply:
> There is no way to enforce that your example can't be written, right?

The example currently compiles from FIR to LLVM, it would seems a hard restriction to restrict all these operations form manipulating function pointers.

>  It sounds like there is "a bug" potentially here if we're saying we're not fully rewriting the function signatures.

It is for sure an oddity. Functions pointers returning arrays sounds like corner cases to me, so I am more inclined of having this slight oddity of still having some function pointers type whose type is not rewritten rather than having to go through every ops that has a "AnyType" result and rewrite the ops.

> I'm wondering if maybe we need to visit all Ops in the module and rewrite FunctionTypes everywhere.

Do you know if MLIR has an infrastructure to rewrite types instead of operation based rewrite pass ? That could help here, though it may make little sense in the general IR sense.

BTW, this function pointer type rewriting issue is generic to all pass that change function signatures, the target rewrite also does not try changing all the function types here and there outside of where it matters:

```
func @takes_func_pointer_with_boxchar(%arg0: (!fir.boxchar<1>) -> f32) {
  %buffer = fir.alloca (!fir.boxchar<1>) -> f32
  fir.store %arg0 to %buffer : !fir.ref<(!fir.boxchar<1>) -> f32>
  return
}
func private @check_the_pass_is_run(!fir.boxchar<1>) -> f32
```
In the output of `fir-opt --target-rewrite`, you can see the ` (!fir.boxchar<1>) -> f32` type is still visible in the function pointers types of `takes_func_pointer_with_boxchar`:
```
  
  func @takes_func_pointer_with_boxchar(%arg0: (!fir.boxchar<1>) -> f32) {
    %0 = fir.alloca (!fir.boxchar<1>) -> f32
    fir.store %arg0 to %0 : !fir.ref<(!fir.boxchar<1>) -> f32>
    return
  }
  func private @check_the_pass_is_run(!fir.ref<!fir.char<1,?>>, i64) -> f32
```

All that to say, I would rather avoid spending too much time here without a functional problem, but we should keep in mind that rewriting function signatures may be more involving than what we are doing if we completely want to erase the previous signature from the IR type system.

Reply:
> Do you know if MLIR has an infrastructure to rewrite types instead of operation based rewrite pass ?
I remember seeing one, but that was a long time ago.

I agree that we shouldn't hold this up at this point. I'll just file this discussion away as something we may have to reconsider later, even if by a small chance.